Fixes related to invertibility of STFT

gauss256 opened the pull request:
**Purpose**

This PR addresses issues related to using `istft` to invert `stft`.

**Overview**

The current code has problems related to the [constant-overlap-add](https://www.dsprelated.com/freebooks/sasp/Overlap_Add_OLA_STFT_Processing.html) (COLA) constraint. The detailed math is available in this [blog post](https://gauss256.github.io/blog/cola.html) but a summary is:

1. The COLA constraint is being enforced by `istft` but that is not necessary and prevents people from using valid windowing configurations.
2. The documentation for `stft` and `istft` suggest that COLA is a necessary condition for invertibility. As the above blog post shows, this is not correct. The necessary and sufficient condition is nonzero-overlap-add (NOLA). This is not a standard term in the field but it is defined in the blog post and in the documentation.
3. A new function `check_NOLA` has been created. It parallels `check_COLA`.
4. Given the availability of `check_NOLA` and because we don't know all the use cases users might have for `istft`, there is now no enforcement of COLA or NOLA in that function.

**Feedback wanted**

1. Is the revised/new documentation of `stft`, `istft`, and `check_COLA` clear and correct?
2. Are the unit tests adequate?
3. The code for `check_NOLA` and `check_COLA` are 95% the same and should probably be refactored. Suggestions for the best way to do this are welcome.
4. Does everyone agree that it is OK to not have checks for COLA or NOLA in `istft`?

Related issue: #8791

tetelestia commented:
The NOLA condition seems to be somewhat trivial. Have you checked to see if the computation time is significant compared to the transform itself? If so, is it worth adding a flag to skip this check?

gauss256 commented:
Some checks failed (for my PR and a couple of others around the same time). The error messages are not related to my PR. 

As suggested in a couple of places, I'm going to close and re-open the PR to force the checks to run again. Hope this doesn't cause any trouble...

tetelestia commented on scipy/signal/spectral.py:
```diff
@@ -1196,10 +1339,6 @@ def istft(Zxx, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
         raise ValueError('noverlap must be less than nperseg.')
     nstep = nperseg - noverlap
 
-    if not check_COLA(window, nperseg, noverlap):
```
It appears to be quite difficult to find a window which fails the NOLA condition, whereas the computation time to check it is not trivial. I think it works better as an external sanity check, as opposed to a forced check.

larsoner replied:
Could we speed it up with:
```
if (window == 0).any():
    <better check>
```

gauss256 commented:
I have now addressed the feedback received so far.

In `istft` the work needed to check the NOLA constraint is already done as part of the normal processing, so I added a line of code to check whether the constraint is satisfied and issue a warning if not. There is now also a unit test specifically for this warning.

larsoner commented on scipy/signal/spectral.py:
```diff
@@ -1246,12 +1387,15 @@ def istft(Zxx, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+
+    # Divide out normalization where non-tiny
+    if np.sum(norm > 1e-10) != len(norm):
+        warnings.warn("NOLA condition failed, STFT may not be invertible")
```
This takes care of my comment above about checking NOLA, right?

gauss256 replied:
Yes

gauss256 commented:
There are still failed tests in `ci/circleci: pypy3`. They don't seem related to my code, but I don't understand the error messages. If someone could let me know if there is still something I need to fix, I would appreciate it.

larsoner commented:
IIRC CircleCI, unlike Travis, does not automatically build a merged version of the PR. So this failure might go away if you rebase.

gauss256 commented:
Rebasing did the trick. Thanks for pointing that out.

e-q commented:
I hope to look it over sometime next week, but that’s not a blocker for merging if deemed necessary. 

> On Jun 12, 2018, at 11:25, Eric Larson <notifications@github.com> wrote:
> 
> @larsoner approved this pull request.
> 
> Rendered docs look good:
> 
> https://7246-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.check_NOLA.html#scipy.signal.check_NOLA
> 
> @e-q do you have time to look?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

larsoner commented:
Yes, feel free to look if you have time @rgommers but I am happy

rgommers commented:
LGTM too, merging. Thanks @gauss256 and @larsoner
